### ISSUE TITLE
General - Current Longterm Plan

### ISSUE THREAD

**Issue by skipjack** [webpack/webpack.js.org#1380]
> Adding this as general outline and todo list for what's planned in terms of high-level changes and organization. Please feel free make suggestions and, if you are interested in picking up something from here, let us know and we can give some guidance. I think it's important that most issues on this repo stay focused on content. 

### Infrastructure

We either need to finish #980 -- which still needs some work, possibly on the [antwar][1] level -- or port the site to a more stable build tool like [phenomic][2] or [gatsby][3]. Porting the site would also offer a few things, like full HMR, that #980 doesn't address but will require some serious thought. Either way these are the infrastructure issues we need to solve:

- [ ] Synchronized routing and `/content` structure with less configuration.
- [ ] Resolve `<title>` issues (e.g. #720). 
- [ ] Display index pages correctly (#1085).
- [ ] Bail if YAML/Markdown parsing fails (#1070).
- [ ] Interactive bundles contain duplicate code (#877).
- [ ] Easier specification of custom layouts per guide (e.g. #1367).
- [ ] Migrate away from _marked_ and clean up process (webpack-contrib/webpack-defaults#73).
- [ ] Better HMR, faster builds, and quicker build tests.
- [ ] Allow illustrations via Mermaid (#469).
- [ ] Turn the site into a [PWA][4] (sync w @TheLarkInn).
- [ ] Consider a better hosting/deployment setup like Netlify (e.g. #845).
- [ ] Mark releases (synced with main package) and archive them (#176, #1017). 
- [ ] Continue component/utility modularization and simplification.

The last one refers to outsourcing much of the custom dev-related stuff to more modular packages like the [voting app][5] or [extracted banner][6].

### Information Architecture (#438)

Certain infrastructure changes should make some of this much easier. Specifically the fact that in newer versions of Antwar as well as in Phenomic and Gatsby -- routing is automatically synchronized/based on the structure of the `/content` directory. This means far less configuration and a more intuitive repository.

- [ ] Top level pages should use a wider layout (e.g. #1367)
- [ ] _Support_, _Development_, and maybe sponsors/backers => _Contribute_ (#1365)
- [ ] _CLI_ nav link should be removed (it will instead just be found under _API_)
- [ ] The articles should then be organized and structured as discussed in #438 

### Content

This is the main reason I've added this umbrella issue and closed many of the smaller dev-related enhancements and bugs. __The main focus of this site should always be the content__. We are making a lot of progress on #1258 but new issues and PRs crop up every day making it very hard to get things done without letting the repo slip back into chaos. This also provides a location for maintainers to quickly touch base and keep track of the overarching plan.

We really need a team of 3 - 5 consistent content writers/maintainers that can help ensure the docs have a fairly consistent voice/approach and each section has well-defined goals. _Please ping me if you are interested_. I've seen many people frustrated with the old docs and parts of the current docs but much fewer who will take the time to help us address these issues once and for all. Don't get me wrong -- we appreciate every one of the [400+ contributors][7] 🎉 🎉 🎉  -- but at the same time it's very hard to build a coherent site without some core authors/reviewers.

Here's where we're at from a high-level:

- [ ] Finish open tickets and missing documentation (or at least keep them under control).
- [ ] Review, organize, and define clear goals for _Concepts_.
- [ ] Review, organize, and define clear goals for _Guides_ (in progress -- #1258).
- [ ] Review, organize, and define clear goals for _Configuration_.
- [ ] Review, organize, and define clear goals for _API_ (including _CLI_ -- sync with @ev1stensberg)

[1]: https://github.com/antwarjs/antwar
[2]: https://phenomic.io/
[3]: https://github.com/gatsbyjs/gatsby
[4]: https://developers.google.com/web/progressive-web-apps/
[5]: https://github.com/webpack-contrib/voting-app
[6]: https://github.com/skipjack/react-banner
[7]: https://github.com/webpack/webpack.js.org/graphs/contributors

**Comment by skipjack** [comment]
Ok... I know some might be saying "woah" with all these issues closed but, as mentioned above I'd like the focus to be finishing the high-priority content issues and then pushing much of this forward. Also many of the issues I closed relate to similar build-process updates whether it be #980 or a port.

I did leave some of the more active threads open like #438 and #469 open... and once work has begun on any of the things above I think it'd be fine to open new issues or re-open the older tickets. My main problem with them is that many are related to one another and have been stagnant for far too long. My hope is that having them in one place will lead more discussion and action to get them resolved.

**Comment by montogeek** [comment]
I could help with the site issues, I have fixed some errors from antwar in the past and also improved a few pages.

**Comment by thescientist13** [comment]
Just saw @TheLarkInn 's tweet, and just wanted to drop a line to say I would love to help out!

**Comment by probablyup** [comment]
Gatsby I think would be a great choice for a different SSG. It addresses several of the items on your checklist simultaneously! https://github.com/gatsbyjs/gatsby/releases/tag/v1.0.1

**Comment by shaodahong** [comment]
Yes, there should be a more concise page, a more readable document and a smoother experience

**Comment by tanyagray** [comment]
Would be interested in helping with this. Haven't been a contributor before so not sure how the time demands would work out. My background is in teaching and in writing training materials for teaching programming in schools, so would offer a keen eye for how docs are interpreted by beginner devs.

**Comment by ev1stensberg** [comment]
For anyone interested in contributing with the CLI, I'm willing to help anyone out on the webpack-cli and how that works. Do you want a summary of plans here or on slack @skipjack ?

**Comment by pspeter3** [comment]
Are you looking for individual contributions at all? I'd be interested in helping with TypeScript + Webpack.

**Comment by TheLarkInn** [comment]
> My background is in teaching and in writing training materials for teaching programming in schools, so would offer a keen eye for how docs are interpreted by beginner devs.

@tanyagray I think you would bring considerable value for "Review, organize, and define clear goals for concepts section" if you'd be willing to tackle that with me. If so break that into a separate issue and tag me, and then we can audit together that entire content section.

**Comment by sushant-j** [comment]
Hi, came here after reading tweet by @TheLarkin.. I would like to contribute but not sure how much can I help as a beginner myself.
I can think of contributing to sections which might be difficult for someone starting out, as I sometimes felt.
Thanks :)

**Comment by ChrisChinchilla** [comment]
I can help with the docs, been looking for a decent open source project to put some time into.

I am a full time technical writer and blogger, and also give presentations to developers about better writing.

This is my site - gregariousmammal.com
And LinkedIn - www.linkedin.com/in/chrischinchilla

Current experience a little out of date, but the past few months I've been helping out with here, PagerDuty, OverOps, Coralogix, CodeShip, Tech Beacon, BCG Ventures and a few others.

I am one of the crazy ones who loves to write docs and spend most of my time reviewing and editing developer submitted documentation, and figuring out what they really mean in a way that everyone can understand.

**Comment by GemmaStiles** [comment]
Hey, I saw the tweet from @TheLarkInn so thought I would drop in and say hello and put my hand up.

I haven't contributed to any open source projects before, but I do love well written docs and consider myself a very detail orientated person.

**Comment by ematipico** [comment]
I follow up by giving my greetings! I'd like to contribute as well, it's been years that I use webpack on production projects, also enterprise level; and I am specialized in it. 

I'd like to spend some time to help to create and fix some documentation :)

Thanks!

**Comment by FabioAntunes** [comment]
Just saw @TheLarkInn tweet, I would love to help. Recently I had to teach PHP developers about all this new front end eco system

**Comment by aendrew** [comment]
Hi! I also noticed @TheLarkInn's tweet. I wrote [a book](https://www.amazon.co.uk/D3-js-4-x-Data-Visualization-Third-ebook/dp/B01MG90SSJ/ref=sr_1_1?ie=UTF8&qid=1499434680&sr=8-1&keywords=rininsland) that involved a fair bit of Webpack config and I'd love to give back by helping out however I can — happy to do writing, editing or any of the site-related stuff (I have experience with both Phenomic and Gatsby).

**Comment by sbaidon** [comment]
Hi! I also saw @TheLarkInn tweet. I have been wanting to contribute to open source and I think this is a great opportunity. I also consider myself a beginner so I don't know if I could be of much help. I am thinking I could help out with translations to spanish if that is in the plans.

Thanks!!

**Comment by geoffdavis92** [comment]
@TheLarkInn @skipjack I commented on twitter that I'd be interested in reviewing content. Though I would be better suited for site updates if those are in the pipeline.

**Comment by CrispyCB** [comment]
@TheLarkInn I was linked this tweet via the Write The Docs slack channel. I don't have discreet technical writing experience, but I've done user guides for everything from PCB manufacture to product testing, so I'd love to help if I'm able.

**Comment by JessRudder** [comment]
Saw the tweet from @TheLarkInn.  I've been looking for an open source project to get involved in more long term and this looks quite promising!  I'm a web developer happy to help write/review docs, work on site issues and/or anything else that would be considered useful.

**Comment by dakebl** [comment]
Hey all, I saw the tweet that @TheLarkInn posted and thought I'd say hello. I'd be happy to help out with the curation effort if you'd like? I have a fair amount of experience in technical documentation from my day job :D

**Comment by mtrivera** [comment]
Hello everyone, I saw the tweet from @TheLarkInn and I'm interested in this project. I contributed to a guide on using Joomla! for administrators.

**Comment by suhasdeshpande** [comment]
As everyone else, I am coming here thanks to @TheLarkInn tweet. I am interested in helping out anyway I can.

**Comment by ev1stensberg** [comment]
One thing we need is a PWA guide along with a `webpack-addons-pwa`. I've spoken a bit with @addyosmani and @TheLarkInn. What I can help you with are: How to create an `webpack-addon` and what the `PWA addon` should look like (programatically), and lastly what the PWA guide should look like. 

We need help on webpack-cli, what I can help and guide you with is as follows (pick one that excites you! :) ): <- That's really a colon

1. Creating an addon and what an PWA addon should look like
2. PWA guide at webpack.js.org or a extended information about the webpack-cli at webpack.js.org
3. General contributions ( see https://github.com/webpack/webpack-cli/issues/87 )


There's various of stuff here for different skillsets: "Beginners" should try out to create an addon, "intermediates" should try option number three ( I will guide you there of course ) and those who want to write docs can do that.

The reason why I used quotes on beginners and intermediates is because these issues don't really require extensive knowledge, anyone could do them. Some issues or points however, on the bullet list; requires more work than the other. I'm happy to guide you and help out with what I can offer and how the CLI is structured. 

Anyone interested?

**Comment by thescientist13** [comment]
@ev1stensberg 
I would be up for working on a PWA guide.  Are you saying that the work there would overlap with contributions **webpack-cli** too?   ( I only ask because working on the CLI sounds pretty rad and so a 2-for-1 would be 🎆 )

**Comment by JessRudder** [comment]
@ev1stensberg I'd love to work on creating the PWA add-on (I'm assuming that's Progressive Web App, let me know if I've assumed wrong).  Would creating a guide be part of that or would someone else be working on the guide while I work on the add-on?

**Comment by ematipico** [comment]
@ev1stensberg I haven't used the webpack CLI so far but I would like to give you some help on that front. I like to deep dive inside the code and give my time as much as I can. I can also help with some documentation, just let me know how I can contribute :)

**Comment by skipjack** [comment]
Wow, thanks to @TheLarkInn for tweeting this out, and thanks to everyone who has volunteered! See the details below and please comment on what you're interested in. Here's what we have and the spots we need to fill...

### Review, organize, and define clear goals for Concepts (#1386)

Sean can speak more to this than I can but essentially these are an alternate starting point to _Guides_ to get a core understanding of webpack right out of the gate. I see the path for webpack developers either going _Concepts_ => _API_ / _Configuration_ => _Guides_ or straight to _Guides_ and then backtracking through the other sections if they want to get started immediately.

- [ ] @tanyagray 
- [ ] @TheLarkInn 
- [ ] Unknown (maybe one more?)

### Review, organize, and define clear goals for Guides (in progress -- #1258)

We're about halfway through the process of organizing and rewriting these guides to ensure they provide consistent examples and logically progress.

- [ ] @TheDutchCoder 
- [ ] @skipjack
- [ ] Unknown (we probably could use one more here)

### Review, organize, and define clear goals for Configuration

_Configuration_ is probably in the best shape of all our content. It's somewhat consistent in terms of formatting and pretty well split up, however a review and a well defined goal couldn't hurt.

- [ ] Unknown
- [ ] Unknown

### Review, organize, and define clear goals for API (including CLI -- sync with @ev1stensberg)

This section has a fair amount of content but 

- [ ] @ev1stensberg 
- [ ] Unknown
- [ ] Unknown

### Finish open tickets and missing documentation

Consider these free range -- pick an existing content related issue, start a small discussion on how it might be tackled, where it would fit best (sometimes this isn't so clear), and get any clarification. Then submit a PR and ping @TheLarkInn @bebraw @simon04 or I (@skipjack) for a review.


<br><p></p><br>

__Note that the first four sections are more long-term. Please take that into consideration.__ The last one, _Finish open tickets..._, is more short term so if you just want to tackle an issue or two that'd be a good fit. Obviously feedback is welcome on any issue or PR, but it helps to have one or two people lead the way in adding and reviewing content so the documentation stays consistent within sections.

Lastly, in terms of porting/fixes/dev work: @montogeek personally I think porting would be the right way to go as gatsby or phenomic tackle other things on that list aside from fixes. @probablyup thanks for that link, I hadn't realized v1 was released and I'm happy to see the new docs site. Once we sort out who is tackling which bits of content let's discuss that a bit more. If you're up for it, I think you (@probablyup), @montogeek, and I could tackle the dev work.

**Comment by ev1stensberg** [comment]
@thescientist13 @JessRudder @ematipico 

Could you all send me an email? Address is in my bio or package.json 👍

**Comment by skipjack** [comment]
@ChrisChinchilla @dakebl seeing as you both have technical writing experience, it would be great if you could grab a section but also help review PRs in general.

@pspeter3 sure we'd be happy to have you improve anything you have time for. If you wanted to submit a PR with improvements to the Typescript guide that would be great! Let's discuss in #1258.

@geoffdavis92 sure will keep that in mind -- I'm not sure how much help we'll need but I'll definitely make sure to include you in the conversation re site updates.

**Comment by leggsimon** [comment]
👋 Hi, I'm Simon, I work at Financial Times. Here from Sean's tweet as well. 

Not sure how much I'll be able to help with the docs but over the last month or so I went from zero knowledge of webpack to having to upgrade all of our stuff to v2. As someone who came in fresh maybe I could help with having a bit simpler descriptions of how things work/what properties do to make it more accessible to webpack beginners.

**Comment by montogeek** [comment]
@skipjack I am up to it, how we choose between both solutions?

**Comment by probablyup** [comment]
I'd definitely be up for rebuilding the existing site via Gatsby! Just let
me know when you'd want that effort to start and I'll start cookin'.

**Comment by skipjack** [comment]
@leggsimon sure is there a specific section you'd be interested in?

**Comment by skipjack** [comment]
@montogeek there's a few things we'll need out of whatever generator we pick. Seeing as gatsby's the most popular and stable, if it hits the points we need then we should probably go that direction. Here's what we need to ensure is doable:

- Markdown customizations (webpack-contrib/webpack-defaults#72)
- Get all page metadata in some data structure that we can use for creating navigations
- Optimize our output by replacing react with preact in production

@bebraw let me know if I missed anything there. @probablyup can you confirm if all those things are doable?

**Comment by probablyup** [comment]
React -> Preact is definitely no problem.

I'll have to look over the documentation a bit more re: metadata. It would
surprise me if that wasn't already built in.

The link for the markdown customization comment looks unrelated?

**Comment by leggsimon** [comment]
@skipjack Certainly one of the things that I had a lot of trouble with was module resolutions with bower_components & node_modules and defining how they're resolved. So maybe that?

**Comment by montogeek** [comment]
@probablyup Maybe this one: https://github.com/webpack-contrib/webpack-defaults/issues/73

**Comment by Jorybraun** [comment]
Ok, I've never contributed to any open source library before could anyone pick a good ticket for me to dedicate my Sunday towards.

**Comment by TheLarkInn** [comment]
@jorybraun et al: i'm going to try and spend 30 minutes today and make a list that might be great for "My first PR"

**Comment by Jorybraun** [comment]
@TheLarkInn : I was thinking about this : but its a big job https://github.com/webpack/webpack.js.org/issues/1397

**Comment by TheLarkInn** [comment]
That would be incredible. I'll tag you there and we can discuss breaking it down into measurable smaller pieces.

**Comment by skipjack** [comment]
@probablyup yep my bad -- the one @montogeek provided is correct (webpack-contrib/webpack-defaults#73) and I updated my comment.

> Certainly one of the things that I had a lot of trouble with was module resolutions with bower_components & node_modules and defining how they're resolved. So maybe that?

@leggsimon sure is there a particular issue or section of the docs you were thinking about addressing in relation to module resolution? Maybe a review of [the configuration page](https://webpack.js.org/configuration/resolve/#resolve) would be a good start?

**Comment by fokusferit** [comment]
@ev1stensberg thanks to @TheLarkInn 's tweet and Lauri Apple's discussion I came across this. I'd be interested in the CLI topic from a long-term perspective. Also trying to get a second developer here from Zalando who is also interested. 

Regarding short term, I'm seeing already some help-wanted tags but not sure if they are a good starting point or already bigger topics ? I've also found an interesting overview here https://github.com/webpack/webpack.js.org/issues/1258#issuecomment-312139191 and maybe the checklist in #1258 there is a good starting point ?

**Comment by ev1stensberg** [comment]
Hi @fokusferit , sweet! I don't have any participation in the short terms you mentioned, but I'm happy to help you with webpack-cli and getting you started there. As slightly mentioned @ematipico is also working with the CLI now. 

If you're available, I can have a hangouts with you and Emanuel to walk you through the codebase? 

If anyone else is interested in figuring out the internals of the CLI, give me a hover ✋

**Comment by skipjack** [comment]
@fokusferit if you could help pick up some of the `help-wanted` tickets that would be awesome. It seems we already have a few volunteers over at #1258 but if there's one you'd really like to tackle feel free to comment.

**Comment by fokusferit** [comment]
@ev1stensberg I will just ping @LeoAref @frontendherodk (hope it works), so we are all interested and up for a Hangout to go through the codebase :) Do you need sth ? Mail ?

@skipjack will go through that, thanks!

**Comment by ev1stensberg** [comment]
@fokusferit shoot me an email, and I'll mail you to arrange a date :)

**Comment by LeoAref** [comment]
@fokusferit @ev1stensberg I'm ready for it 👍

**Comment by skipjack** [comment]
@ChrisChinchilla @dakebl I know this thread has gotten a bit noisy but if either or you are still interested please don't hesitate to reach out to me either via gitter or by email.

**Comment by ev1stensberg** [comment]
Hangouts is next Tuesday 2 - 3 pm or 3 - 4 pm (Berlin-time), I'll post a link here a couple of hours before we start 🎆

**Comment by MoOx** [comment]
If you have any question about Phenomic, feel free to poke me. v1 alpha is still early (note: not in term of API stability because we learned from our 0.x ^^) but we would like to help you if possible as we think Phenomic is a perfect fit for webpack.js.org.

**Comment by frontendherodk** [comment]
@ev1stensberg Sounds good!

**Comment by fokusferit** [comment]
To everyone who is interested as we have some people also from Zalando interested in the topic:

- We organized the hangout regarding webpack CLI today from 3:30 - 4:30 pm (CEST, Central European Time), you can join us later [this link](https://hangouts.google.com/hangouts/_/zalando.de/webpack-cli) see you soon!

**Comment by ev1stensberg** [comment]
This is also a CLI meeting about future plans :)

**Comment by ChrisChinchilla** [comment]
@skipjack @dakebl I completely missed notifications on this post, but still keen and happy to coordinate a meeting. I am in CEST.

**Comment by skipjack** [comment]
@ChrisChinchilla yeah are you on Gitter? @dakebl you? We could set up a little room there to discuss at first or just talk in the webpack/docs room.

**Comment by skipjack** [comment]
@MoOx will do, thanks for the offer! I think what I'll do is close this issue out once we finish finding content writers and open a separate issue specific to development work. I'll make sure to cc you.

**Comment by ChrisChinchilla** [comment]
@skipjack @dakebl I am now…

**Comment by skipjack** [comment]
Ok I'm going to close this out this weekend and open a dev-specific issue/checklist in its place (as I mentioned above). If anyone is still interested in contributing but hasn't found a section or issue, please don't hesitate to ping me. What we be great is if we could find one or two people to lead the content work on the `/api` section.

**Comment by skipjack** [comment]
Closing in favor of more targeted high-level issues (e.g. #1258, #1386, and #1525) now that we've started to get a bunch of help.

If anyone from this thread has not found an issue to tackle or section to dive into please don't hesitate to ping me on gitter and I will gladly help you with that. This site is finally under (just barely) one page of issues and we're going to try to keep it that way. Any and all help is much appreciated!

@thescientist13  @JessRudder if either or both of you wanted to tackle #1145 (writing PWA with webpack guide) that would be fantastic.
